src/python/WMCore/Services/WMStats/WMStatsReader.py

```diff
@@ -304,6 +304,15 @@ def getT0ActiveData(self, jobInfoFlag=False):
 
         return self.getRequestByStatus(T0_ACTIVE_STATUS, jobInfoFlag)
 
+    def getT0ArchivedData(self, jobInfoFlag=False, startTime=0):
+        """
+        A wrapper method to call getRequestByStatusAndStartTime (To be used mostly by T0)
+        :param jobInfoFlag: A flag to state if to update the workflow with jobInfo too if set it implies `detail=True` as well.
+        :param startTime:   unix start timestamp
+        :return:            A dictionary with worklflows info in status *-archived since startTime.
+        """
+        return self.getRequestByStatusAndStartTime(ARCHIVED_STATUS, detail=True, jobInfoFlag=jobInfoFlag, startTime=startTime)
```

Comment:
This is exactly what we discussed in the past. Querying for archived workflows is bad. You are basically asking for all the data that we have in the system, starting in 2010 or so.

To follow up on this comment, please provide:
a) the exact status for T0 archived workflows (they definitely do not use aborted-archived and rejected-archived
b) and some evaluation of how such query behaves with the current data (amount of data returned over the network, amount of data loaded in memory on the client side and total time to complete this call).

Reply:
ok, I was having troubles making it work perfectly with `wmstatsReader`, but fetching these from the `centralreqDB` directly using the very same method gives the following statistics.
If we fetch all details for all the workflows in status `normal-archived` since period of time `X` from T0 production `reqmon`, where `X` is gradually incremented:
```
In [1]: from Utils.GetStats import *; import time; from pprint import pformat

In [2]: days=15
   ...: while days < 4*365:
   ...:     days = days*2; startTime = int(time.time()) - days*24*60*60
   ...:     print("------------------------------------------------------------")
   ...:     print("Period: now - %s days." % days)
   ...:     profResult = getStatsReqDB("normal-archived", startTime=startTime, detail=True)
   ...:     print("Runtime: %s sec." % profResult[0])
   ...:     print("============================================================\n\n")
   ...: 
------------------------------------------------------------
Period: now - 30 days.
centralRequestDBURL: https://cmsweb.cern.ch/couchdb/t0_request
NumRecords: 2041

Filename: /data/tmp/WMCore.venv3/srv/WMCore/src/python/Utils/GetStats.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    19     48.3 MiB     48.3 MiB           1   @timeFunction
    20                                         @profile
    21                                         def getStatsReqDB(status, startTime=0, detail=False):
    22     58.2 MiB      9.8 MiB           1       wflows = centralRequestDBReader.getRequestByStatusAndStartTime(status, startTime=startTime, detail=detail)
    23     58.2 MiB      0.0 MiB           1       print("centralRequestDBURL: %s\nNumRecords: %s\n"  % (centralRequestDBURL, len(wflows)))


Runtime: 4.5563 sec.
============================================================


------------------------------------------------------------
Period: now - 60 days.
centralRequestDBURL: https://cmsweb.cern.ch/couchdb/t0_request
NumRecords: 4306

Filename: /data/tmp/WMCore.venv3/srv/WMCore/src/python/Utils/GetStats.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    19     52.4 MiB     52.4 MiB           1   @timeFunction
    20                                         @profile
    21                                         def getStatsReqDB(status, startTime=0, detail=False):
    22     66.2 MiB     13.8 MiB           1       wflows = centralRequestDBReader.getRequestByStatusAndStartTime(status, startTime=startTime, detail=detail)
    23     66.2 MiB      0.0 MiB           1       print("centralRequestDBURL: %s\nNumRecords: %s\n"  % (centralRequestDBURL, len(wflows)))


Runtime: 8.392 sec.
============================================================


------------------------------------------------------------
Period: now - 120 days.
centralRequestDBURL: https://cmsweb.cern.ch/couchdb/t0_request
NumRecords: 5422

Filename: /data/tmp/WMCore.venv3/srv/WMCore/src/python/Utils/GetStats.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    19     54.0 MiB     54.0 MiB           1   @timeFunction
    20                                         @profile
    21                                         def getStatsReqDB(status, startTime=0, detail=False):
    22     70.8 MiB     16.9 MiB           1       wflows = centralRequestDBReader.getRequestByStatusAndStartTime(status, startTime=startTime, detail=detail)
    23     70.8 MiB      0.0 MiB           1       print("centralRequestDBURL: %s\nNumRecords: %s\n"  % (centralRequestDBURL, len(wflows)))


Runtime: 10.9677 sec.
============================================================


------------------------------------------------------------
Period: now - 240 days.
centralRequestDBURL: https://cmsweb.cern.ch/couchdb/t0_request
NumRecords: 13357

Filename: /data/tmp/WMCore.venv3/srv/WMCore/src/python/Utils/GetStats.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    19     55.1 MiB     55.1 MiB           1   @timeFunction
    20                                         @profile
    21                                         def getStatsReqDB(status, startTime=0, detail=False):
    22     95.6 MiB     40.5 MiB           1       wflows = centralRequestDBReader.getRequestByStatusAndStartTime(status, startTime=startTime, detail=detail)
    23     95.6 MiB      0.0 MiB           1       print("centralRequestDBURL: %s\nNumRecords: %s\n"  % (centralRequestDBURL, len(wflows)))


Runtime: 26.9948 sec.
============================================================


------------------------------------------------------------
Period: now - 480 days.
centralRequestDBURL: https://cmsweb.cern.ch/couchdb/t0_request
NumRecords: 16615

Filename: /data/tmp/WMCore.venv3/srv/WMCore/src/python/Utils/GetStats.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    19     55.6 MiB     55.6 MiB           1   @timeFunction
    20                                         @profile
    21                                         def getStatsReqDB(status, startTime=0, detail=False):
    22    112.6 MiB     57.0 MiB           1       wflows = centralRequestDBReader.getRequestByStatusAndStartTime(status, startTime=startTime, detail=detail)
    23    112.6 MiB      0.0 MiB           1       print("centralRequestDBURL: %s\nNumRecords: %s\n"  % (centralRequestDBURL, len(wflows)))


Runtime: 42.1017 sec.
============================================================


------------------------------------------------------------
Period: now - 960 days.
centralRequestDBURL: https://cmsweb.cern.ch/couchdb/t0_request
NumRecords: 19944

Filename: /data/tmp/WMCore.venv3/srv/WMCore/src/python/Utils/GetStats.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    19     62.4 MiB     62.4 MiB           1   @timeFunction
    20                                         @profile
    21                                         def getStatsReqDB(status, startTime=0, detail=False):
    22    128.0 MiB     65.6 MiB           1       wflows = centralRequestDBReader.getRequestByStatusAndStartTime(status, startTime=startTime, detail=detail)
    23    128.0 MiB      0.0 MiB           1       print("centralRequestDBURL: %s\nNumRecords: %s\n"  % (centralRequestDBURL, len(wflows)))


Runtime: 40.6828 sec.
============================================================


------------------------------------------------------------
Period: now - 1920 days.
....
ERROR:
error: (28, 'Operation timed out after 270000 milliseconds with 64569481 out of -1 bytes received')

```

Reply:
While if we just fetch RequestNames, without details, the picture gets pretty good:
```
In [3]: days=15
   ...: while days < 4*365:
   ...:     days = days*2; startTime = int(time.time()) - days*24*60*60
   ...:     print("------------------------------------------------------------")
   ...:     print("Period: now - %s days." % days)
   ...:     profResult = getStatsReqDB("normal-archived", startTime=startTime, detail=False)
   ...:     print("Runtime: %s sec." % profResult[0])
   ...:     print("============================================================\n\n")
   ...: 
------------------------------------------------------------
Period: now - 30 days.
centralRequestDBURL: https://cmsweb.cern.ch/couchdb/t0_request
NumRecords: 2041

Filename: /data/tmp/WMCore.venv3/srv/WMCore/src/python/Utils/GetStats.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    19    130.3 MiB    130.3 MiB           1   @timeFunction
    20                                         @profile
    21                                         def getStatsReqDB(status, startTime=0, detail=False):
    22    130.3 MiB      0.0 MiB           1       wflows = centralRequestDBReader.getRequestByStatusAndStartTime(status, startTime=startTime, detail=detail)
    23    130.3 MiB      0.0 MiB           1       print("centralRequestDBURL: %s\nNumRecords: %s\n"  % (centralRequestDBURL, len(wflows)))


Runtime: 0.1358 sec.
============================================================


------------------------------------------------------------
Period: now - 60 days.
centralRequestDBURL: https://cmsweb.cern.ch/couchdb/t0_request
NumRecords: 4293

Filename: /data/tmp/WMCore.venv3/srv/WMCore/src/python/Utils/GetStats.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    19    130.3 MiB    130.3 MiB           1   @timeFunction
    20                                         @profile
    21                                         def getStatsReqDB(status, startTime=0, detail=False):
    22    131.1 MiB      0.8 MiB           1       wflows = centralRequestDBReader.getRequestByStatusAndStartTime(status, startTime=startTime, detail=detail)
    23    131.1 MiB      0.0 MiB           1       print("centralRequestDBURL: %s\nNumRecords: %s\n"  % (centralRequestDBURL, len(wflows)))


Runtime: 0.3183 sec.
============================================================


------------------------------------------------------------
Period: now - 120 days.
centralRequestDBURL: https://cmsweb.cern.ch/couchdb/t0_request
NumRecords: 5422

Filename: /data/tmp/WMCore.venv3/srv/WMCore/src/python/Utils/GetStats.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    19    130.4 MiB    130.4 MiB           1   @timeFunction
    20                                         @profile
    21                                         def getStatsReqDB(status, startTime=0, detail=False):
    22    131.7 MiB      1.3 MiB           1       wflows = centralRequestDBReader.getRequestByStatusAndStartTime(status, startTime=startTime, detail=detail)
    23    131.7 MiB      0.0 MiB           1       print("centralRequestDBURL: %s\nNumRecords: %s\n"  % (centralRequestDBURL, len(wflows)))


Runtime: 0.6655 sec.
============================================================


------------------------------------------------------------
Period: now - 240 days.
centralRequestDBURL: https://cmsweb.cern.ch/couchdb/t0_request
NumRecords: 13357

Filename: /data/tmp/WMCore.venv3/srv/WMCore/src/python/Utils/GetStats.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    19    130.7 MiB    130.7 MiB           1   @timeFunction
    20                                         @profile
    21                                         def getStatsReqDB(status, startTime=0, detail=False):
    22    135.6 MiB      4.9 MiB           1       wflows = centralRequestDBReader.getRequestByStatusAndStartTime(status, startTime=startTime, detail=detail)
    23    135.6 MiB      0.0 MiB           1       print("centralRequestDBURL: %s\nNumRecords: %s\n"  % (centralRequestDBURL, len(wflows)))


Runtime: 0.7074 sec.
============================================================


------------------------------------------------------------
Period: now - 480 days.
centralRequestDBURL: https://cmsweb.cern.ch/couchdb/t0_request
NumRecords: 16615

Filename: /data/tmp/WMCore.venv3/srv/WMCore/src/python/Utils/GetStats.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    19    130.7 MiB    130.7 MiB           1   @timeFunction
    20                                         @profile
    21                                         def getStatsReqDB(status, startTime=0, detail=False):
    22    136.9 MiB      6.2 MiB           1       wflows = centralRequestDBReader.getRequestByStatusAndStartTime(status, startTime=startTime, detail=detail)
    23    136.9 MiB      0.0 MiB           1       print("centralRequestDBURL: %s\nNumRecords: %s\n"  % (centralRequestDBURL, len(wflows)))


Runtime: 0.8415 sec.
============================================================


------------------------------------------------------------
Period: now - 960 days.
centralRequestDBURL: https://cmsweb.cern.ch/couchdb/t0_request
NumRecords: 19944

Filename: /data/tmp/WMCore.venv3/srv/WMCore/src/python/Utils/GetStats.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    19    130.9 MiB    130.9 MiB           1   @timeFunction
    20                                         @profile
    21                                         def getStatsReqDB(status, startTime=0, detail=False):
    22    138.6 MiB      7.7 MiB           1       wflows = centralRequestDBReader.getRequestByStatusAndStartTime(status, startTime=startTime, detail=detail)
    23    138.6 MiB      0.0 MiB           1       print("centralRequestDBURL: %s\nNumRecords: %s\n"  % (centralRequestDBURL, len(wflows)))


Runtime: 0.8592 sec.
============================================================


------------------------------------------------------------
Period: now - 1920 days.
centralRequestDBURL: https://cmsweb.cern.ch/couchdb/t0_request
NumRecords: 217042

Filename: /data/tmp/WMCore.venv3/srv/WMCore/src/python/Utils/GetStats.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    19    131.0 MiB    131.0 MiB           1   @timeFunction
    20                                         @profile
    21                                         def getStatsReqDB(status, startTime=0, detail=False):
    22    246.8 MiB    115.7 MiB           1       wflows = centralRequestDBReader.getRequestByStatusAndStartTime(status, startTime=startTime, detail=detail)
    23    246.8 MiB      0.0 MiB           1       print("centralRequestDBURL: %s\nNumRecords: %s\n"  % (centralRequestDBURL, len(wflows)))


Runtime: 10.3835 sec.
============================================================

```

Reply:
Something is wrong with these stats, given that "all details" calls have a smaller footprint compared to the "request names only".

Anyhow, I would suggest to put these checks/profiling on hold until you exhaust all the possibilities of decoupling rucio rule deletion from the workflow archival process (as we discussed in the WMCore chat today).